ui/components-react/src/components-react/filter-builder/FilterBuilderState.ts

```diff
@@ -167,34 +170,163 @@ export class PropertyFilterBuilderActions {
+        }
+      };
+
+      setErrorMessages(state.rootGroup);
+    });
+  }
 }
 
 /**
  * Function to check if supplied [[PropertyFilterBuilderRuleGroupItem]] is [[PropertyFilterBuilderRuleGroup]].
- * @internal
+ * @beta
  */
 export function isPropertyFilterBuilderRuleGroup(
   item: PropertyFilterBuilderRuleGroupItem
 ): item is PropertyFilterBuilderRuleGroup {
   return (item as any).items !== undefined;
 }
 
+/**
+ * Props for [[usePropertyFilterBuilder]]
+ * @beta
+ */
+export interface UsePropertyFilterBuilderProps {
+  /** Initial filter for [[PropertyFilterBuilder]] */
+  initialFilter?: PropertyFilter;
+  /** Custom rule validator to be used when [[buildFilter]] is invoked. Should return error message or undefined, if rule is valid. */
+  ruleValidator?: (rule: PropertyFilterBuilderRule) => string | undefined;
+}
+
+/**
+ * Type for [[usePropertyFilterBuilder]] return object.
+ * @beta
+ */
+export interface UsePropertyFilterBuilderValues {
+  /** Root group of the [[PropertyFilterBuilder]]. */
+  rootGroup: PropertyFilterBuilderRuleGroup;
+  /** Actions for manipulating [[PropertyFilterBuilder]] state. */
+  actions: PropertyFilterBuilderActions;
+  /**
+   * Filter builder for building [[PropertyFilter]] from [[PropertyFilterBuilder]] state.
+   * @returns PropertyFilter if all rules are valid, otherwise sets error messages fow invalid rules using
+   * default rule validator or custom validator provided from [[UsePropertyFilterBuilderProps]] and returns undefined.
+   */
+  buildFilter: () => string | undefined;
+}
+
 /**
  * Custom hook that creates state for [[PropertyFilterBuilder]] component. It creates empty state or initializes
  * state from supplied initial filter.
- * @internal
+ * @beta
  */
-export function usePropertyFilterBuilderState(initialFilter?: PropertyFilter) {
-  const [state, setState] = React.useState<PropertyFilterBuilderState>(() =>
+export function usePropertyFilterBuilder(
+  props?: UsePropertyFilterBuilderProps
+) {
+  const { initialFilter, ruleValidator } = props ?? {
+    initialFilter: undefined,
+    ruleValidator: undefined,
+  };
+  const [state, setState] = React.useState<PropertyFilterBuilderState>(
     initialFilter
       ? convertFilterToState(initialFilter)
       : { rootGroup: createEmptyRuleGroup() }
   );
+
   const [actions] = React.useState(
     () => new PropertyFilterBuilderActions(setState)
   );
 
-  return { state, actions };
+  const buildFilter = () => {
+    const ruleIdsAndErrorMessages = ruleGroupItemValidator({
+      item: state.rootGroup,
+      ruleValidator,
+    });
+    if (ruleIdsAndErrorMessages.size === 0) {
+      actions.setRuleErrorMessages(new Map<string, string>());
+      return buildPropertyFilter(state.rootGroup);
+    } else {
+      actions.setRuleErrorMessages(ruleIdsAndErrorMessages);
+      return undefined;
+    }
+  };
+  return { rootGroup: state.rootGroup, actions, buildFilter };
+}
+
+interface RuleGroupItemValidatorProps {
```

Comment:
I think this interface can be removed and `ruleGroupItemValidator` can take `item` and `ruleValidator` as separate parameters.

Reply:
fixed